### ISSUE TITLE
[Question] Combining QAT and Sparsity Training

### ISSUE THREAD

**Issue by CaptainDario** [pytorch/ao#2310]
First of all, thank you for all the time and effort invested in this project to make (large) models more accessible.
I am fairly new to optimizing my models using sparsity, and therefore, wanted to ask if my understanding of this library is correct.
In general, I would like to train my model using sparsity and QAT.

For QAT, I would follow this [guide](https://github.com/pytorch/ao/blob/main/torchao/quantization/qat/README.md#quantize_-api-recommended).
Now I am curious how to correctly use this together with sparsity.
I assume this `swap_linear_with_semi_sparse_linear(model, sparse_config)` is the correct snippet ([guide](https://github.com/pytorch/ao/tree/main/torchao/sparsity/training#quickstart)).
If I want to combine these two optimizations, what is the correct way to do so?

1. Train baseline
2. Train a sparse model
3. Train a sparse and quantization-aware model

Additionally, I found this statement

> A fully sparse 2:4 trained model exhibited a -0.5 pp accuracy drop; we were able to further reduce the accuracy loss to -0.1 pp by first training with 2:4 sparsity enabled and then switching over to normal dense training.

Does this mean adding a step?

4. Revert sparsity using `swap_semi_sparse_linear_with_linear(model)` and train

Lastly, the sparsity `sparsify_ ` and quantization `quantize_` need to be 'applied'.

I would greatly appreciate your input on this.

**Comment by supriyar** [comment]
@CaptainDario the QAT process is typically applied during fine-tuning and the sparse training work was built for pre-training. I don't think as built these can two be composed and would require additional work to make that happen. Is your goal to get a sparse quantized model for inference or something else?

cc @andrewor14 @jcaip for further comments

**Comment by dariyooo** [comment]
Thank you for your reply.

Yes, I am trying to build a faster/smaller model for inference, and for this, I am exploring the current state of options.
If I first train the model using sparsity, will it be lost when training again using QAT?
Is it possible to use Torch's inbuilt pruning utilities, such as `global_unstructured`, to first prune the model and then retrain using ao's QAT?

**Comment by andrewor14** [comment]
Hi @CaptainDario, these two features were not designed to compose today and so there will be extra work that needs to be done. If you do want to hack something together and use the features in the same training loop, you could modify `SemiSparseLinear` to apply a `FakeQuantizer` to the weights and activations, similar to what we do for `FakeQuantizerLinear` [here](https://github.com/pytorch/ao/blob/12398428a0573374c3971d9cb66d47185a11feff/torchao/quantization/qat/linear.py#L105).Alternatively, I think it would make more sense to separate the two flows into different pre-training and fine-tuning steps as Supriya has mentioned. But we've never tested this and so it's unclear how this will perform. Also @jcaip can comment on `pytorch/pytorch`'s in-built pruning utilities but I don't think those are well maintained. The most actively developed sparsity APIs should be the ones in torchao.

**Comment by jcaip** [comment]
I would +1 @andrewor14 suggestion to do this in a piecewise manner, first applying semi-structured sparse training and then QAT afterwards. 

> If I first train the model using sparsity, will it be lost when training again using QAT?

Yes, if you convert back to `nn.Linear` the QAT training will make the model non-sparse. You can attach a `WeightNormPruner` to mask the weights in the forwards pass (but note that this will not be accelerated).  You can read some more about this [here](https://docs.pytorch.org/ao/stable/sparsity.html)
```
import torch
from torch.sparse import to_sparse_semi_structured, SparseSemiStructuredTensor

from torchao.prototype.sparsity.sparsifier.weight_norm_sparsifier import WeightNormSparsifier


# bfloat16 CUDA model
model = model.half().cuda()

# Accuracy: Finding a sparse subnetwork
sparse_config = []
for name, mod in model.named_modules():
   if isinstance(mod, torch.nn.Linear):
      sparse_config.append({"tensor_fqn": f"{name}.weight"})

sparsifier = WeightNormSparsifier(sparsity_level=1.0,
                                 sparse_block_shape=(1,4),
                                 zeros_per_block=2)

# attach FakeSparsity
sparsifier.prepare(model, sparse_config)
sparsifier.step()

# train your model here
```

You can then accelerate the subsequent model with `sparsify_ / quantize_` by following the steps in the README [here](https://github.com/pytorch/ao/tree/main/torchao/sparsity#int8-dynamic-quant--24-sparasity). 

`torch.nn.utils.prune` is currrently not under active development, it may be possible to do the same thing but I would recommend using the ao.pruning flow.

**Comment by dariyooo** [comment]
Thank you very much!

So basically, the flow should be

1. Train (pre-warmed) model with desired sparsity using `WeightNormSparsifier`
2. Apply `WeightNormPruner` and fine-tune using QAT
3. Accelerate for inference using `sparsify_ + quantize_`

?

**Comment by jcaip** [comment]
Yeah that looks right, but a couple notes

> Apply WeightNormPruner and fine-tune using QAT

You don't need to apply both `WeightNormPruner` and `WeightNormSparsifier`. 

Between step 2 and 3, you'll need to call `sparsifier.squash_mask()` to fuse the masks into the weights, you can't do this before QAT training because then the sparsity will be lost.

**Comment by dariyooo** [comment]
Okay, thank you!
I will close this issue and try this in the coming days.
If I encounter problems, I will reopen this issue.
